### ISSUE TITLE
Null coalescing operator for strings

### ISSUE THREAD

**Issue by vsfeedback** [dotnet/roslyn#31315]
<p>While it's possible to use the null coalescing operator with strings to assign a default value when <b>the strings can be null</b>, the null coalescing operator can't be use with TextBox.Text properties because they are never null, but they can be empty strings.</p><p>It would be helpful to extend the null coalescing operator for strings and make them treating empty string as they were null strings. By this way instead of programming:</p><pre>s0 = String.IsNullOrEmpty(myTextBox.Text) ? sDefault : myTextBox.Text;</pre><p></p><p>we can use the shorter and easier to read form:</p><pre>s0 = myTextBox.Text ?? sDefault :</pre>

_This issue has been moved from https://developercommunity.visualstudio.com/content/idea/374856/null-coalescing-operator-for-strings.html
VSTS ticketId: 723588_
_These are the original issue comments:_

Fiona Niu[MSFT] on 11/4/2018, 08:13 PM (17 days ago): <p>Thank you for taking the time to provide your suggestion. We will do some preliminary checks to make sure we can proceed further.&nbsp;You will hear from us in about a week on our next steps. </p>

**Comment by Neme12** [comment]
4. Is it really the case that both null and empty string are potential values of your string with different semantics, which you want to ignore and treat them the same?

**Comment by AngeloMascaro** [comment]
Neme12 you are right: they could have a different meaning. I normally treat them as the same, but other developers could give them a different meaning.

**Comment by TheFanatr** [comment]
This would also be a major breaking change and one that isn't strictly obvious. It is very common that when someone uses the expression `stringOne ?? stringTwo`, where `stringOne` could an empty and `stringTwo` could simply be a fallback if somehow `stringOne` is null, it is actually expected that the fallback not be used when `stringOne` is simply empty or whitespace. So if this behaviour change was applied to that expression, then it would fail to fulfill it's original purpose thus creating a bug, but to add to that, there is no way to opt-out of this functionality so then an expression such as `stringOne is null ? stringTwo : stringOne`, which creates the exact same problem that you are trying to solve, but for other scenarios.

**Comment by BhaaLseN** [comment]
In terms of business domain,  `null` and `string.Empty` are generally considered to be the same - ie. _the textbox is empty to the end user_ (regardless of whether we have `null` or `string.Empty` in there).

Technically there is an obvious difference, and I'm pretty sure we'll be stuck with that. Too much potential for breakage in existing code.

And as Cyrus mentioned, you can easily write an extension method for that. In fact, we have one that takes a `params string[]` and returns the first that isn't `string.IsNullOrEmpty`.

**Comment by AngeloMascaro** [comment]
Yes, I understand the impact on the existing code in case of a change. In effect I use quite often extension methods, but in this case I didn't consider this possibilty. Thanks for the suggestion.

**Comment by Neme12** [comment]
Correction: in terms of business domain, usually only one of those is a valid value (probably an empty string) but developers get sloppy with their contracts.
